test/server/lib/SamlConfig.ts

```diff
@@ -0,0 +1,202 @@
+          SAMLResponse: loginSamlPayload.toString(),
+          RelayState: relayState,
+        }),
+      });
+      assert.equal(samlResp.status, 302, 'should redirect');
+      const spUrl = new URL(samlResp.headers.get('location') || '');
+      assert.equal(spUrl.pathname, '/', 'should redirect to main');
+
+      // Finish following the redirect
+      await fetch(homeUrl(), {headers});
+
+      // Let's check the user was created via SAML
+      jordi = await getDbManager().getExistingUserByLogin('jordi@getgrist.com');
+      assert.notEqual(jordi, undefined, 'jordi should exist');
+      assert.notEqual(jordi?.firstLoginAt, null, 'jordi should have logged in');
+
+      // Finally, we log out
+      const logoutResp = await fetch(homeUrl() + "/o/docs/logout", {
+        redirect: 'manual',
+        headers
+      });
+      const idpLogoutUrl = new URL(logoutResp.headers.get('location') || '');
+      assert.equal(idpLogoutUrl.origin + idpLogoutUrl.pathname, idpHost, 'should redirect to IdP host');
+      const samlLogoutRequest = decodeSaml(idpLogoutUrl.searchParams.get('SAMLRequest') || '');
+      assert.match(samlLogoutRequest, /samlp:LogoutRequest/,
+                   'SAML logout request should have the right root element');
+
+      const logoutRelayState = idpLogoutUrl.searchParams.get('RelayState') || '';
+      assert.match(logoutRelayState, /permit-external-[0-9a-f-]+/, 'should have a relay state');
+
+      // Pretend that the IdP server sent back a logout response
+      const logoutSamlPayload = await fse.readFile(logoutSamlPath);
+      const samlLogoutResp = await fetch(samlUrl("/assert"), {
+        redirect: 'manual',
+        method: "POST",
+        body: new URLSearchParams({
+          SAMLResponse: logoutSamlPayload.toString(),
+          RelayState: logoutRelayState
+        })
+      });
+      assert.equal(samlLogoutResp.status, 302, 'should redirect');
+      const spLogoutUrl = new URL(samlLogoutResp.headers.get('location') || '');
+      assert.equal(spLogoutUrl.pathname, '/o/docs/signed-out', 'should redirect to main');
+
+      // Finish the logout redirect
+      await fetch(homeUrl() + '/o/docs/signed-out', {headers});
+    });
+
+    it('should allow IdP-initiated logins', async () => {
+      // Ensure a clean start
+      let jordi = await getDbManager().getExistingUserByLogin('jordi@getgrist.com');
+      assert.equal(jordi, undefined, 'jordi should not exist');
+
+      // Grist, our SP, is sitting there quietly, minding its own
+      // business, when suddenly AN UNSOLICITED SAML LOGIN APPEARS.
+      const loginSamlPayload = await fse.readFile(loginSamlPath);
+      const samlResp = await fetch(samlUrl("/assert"), {
+        redirect: 'manual',
+        method: "POST",
+        credentials: 'include',
+        body: new URLSearchParams({
+          SAMLResponse: loginSamlPayload.toString(),
+          // No RelayState, the SAML comes unrequested, nothing to relay
```

Comment:
It would be nice to test it also with RelayState, a valid URL and an invalid one, to exercise the new code that checks those.